### PR TITLE
Chunked message receiving

### PR DESCRIPTION
**Pull request by miwee**
handle chunked packet receiving case. a packet may come as split into 2 or more parts depending on network latency and data publish rate. the issue is more prominent if data publish rate is very high.

### CODE REVIEW

**Comment by suvash** [comment]
Thanks for the PRs.👏
Will take a proper look at this after I'm back in a week.

**Comment by suvash** [comment]
This is a fairly big change. After a quick look, I see that the interface for decoder's reply has changed, which is probably something that had to be done at some point.
Once again, I'd like more tests around this patch, as this is clearly adding functionality, not just a cleanup.
Also, I need to refresh my memory of the mqtt spec and investigate why the existing `:ok` test expectations were updated to be `:error`. Maybe you have an explanation for this @miwee ?

**Comment by miwee** [comment]
I admit that :error is not a good way to indicate the situation, as it is no error as per standard mqtt spec. It should be more like :partial or :wait_for_more. The important point is, it must be different than :ok.

The existing implementation assumes that complete mqtt packet is received in one single chunk like <packet0><packet1><packet2><packet3>.. But there is a edge case when packets are sent at high speed (I tried with 10000 packets per second), the situation becomes something like this <packet0_start><packet0_rest+packet1_start><packet1_rest+packet2+packet3_start>...... packetx_start and packetx_rest refer here just two parts of packet without any specific marking. 

To put it another way, if you received a packet, which is partial and missing last few bytes, what will you do? The existing implementation crashes for current packet (and continues for next packet as usual in true erlang style). However better approach is to check while decoding length, whether or not we have complete packet. In case we don't have complete packet, store the data and wait for next set of data to arrive before doing any further processing.

**Comment by suvash** [comment]
@miwee Thanks for the explanatation. I agree with most of your points. Can you share the bit of code that you used to stress test this lib. I'd like to be able to write tests for the issue or at least be able to reproduce it during stress tests, before I start working on this.

**Comment by miwee** [comment]
The actual code where I encountered this issue was part of larger application. But I just tried this minimal python script, and it reproduced the bug. the master branch crashes, while subscribed on topic "fast_publish_topic". where as the patched branch keep receiving all the messages. The script is:

```
import paho.mqtt.client as mqtt

client = mqtt.Client(client_id="fast_publish_client")

client.connect("127.0.0.1", 1883, 120)
client.loop_start()

i = 0
while True:
  i += 1
  client.publish("fast_publish_topic", "test data %d" % i)
```

**Comment by suvash** [comment]
squashed and used in #28 Landed in master now. Thanks for the PR 🍻
